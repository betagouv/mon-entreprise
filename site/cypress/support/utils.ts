export const fr = Cypress.env('language') === 'fr'
export const baseUrl = Cypress.config('baseUrl')
