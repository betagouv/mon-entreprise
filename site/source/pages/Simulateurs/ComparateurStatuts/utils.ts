export type ValueType =
	| string
	| number
	| boolean
	| null
	| Record<string, unknown>
export type BestOption = {
	type: 'sasu' | 'ei' | 'ae'
	value?: ValueType
}
export const getBestOption = (options: BestOption[]) => {
	const sortedOptions = options.sort(
		(option1: BestOption, option2: BestOption) => {
			if (option1.value === null || option1.value === undefined) {
				return 1
			}
			if (option2.value === null || option2.value === undefined) {
				return -1
			}

			if (option1.value === option2.value) return 0
			// console.log(option1.value, option2.value, option1.value > option2.value)

			return option1.value > option2.value ? -1 : 1
		}
	)

	return sortedOptions?.[0]?.type
}
