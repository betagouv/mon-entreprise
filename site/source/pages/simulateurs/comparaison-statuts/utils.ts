import { DottedName } from 'modele-social'
import Engine, { Evaluation } from 'publicodes'

import { StatutType } from '@/components/StatutTag'

export type OptionType = {
	engine: Engine<DottedName>
	name: StatutType
	value: Evaluation
}
export const getBestOption = (options: OptionType[], leastIsBest: boolean) => {
	const sortedOptions = [...options].sort(
		(option1: OptionType, option2: OptionType) => {
			if (option1.value === null || option1.value === undefined) {
				return leastIsBest ? -1 : 1
			}
			if (option2.value === null || option2.value === undefined) {
				return leastIsBest ? 1 : -1
			}

			if (option1.value === option2.value) return 0

			return (option1.value > option2.value ? -1 : 1) * (leastIsBest ? -1 : 1)
		}
	)

	return sortedOptions?.[0]?.name
}
