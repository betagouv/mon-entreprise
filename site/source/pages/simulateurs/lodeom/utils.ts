import { sumAll } from 'effect/Number'
import { DottedName } from 'modele-social'
import Engine from 'publicodes'

import { Situation } from '@/domaine/Situation'

// TODO: remplacer "salarié . cotisations . assiette" par "salarié . rémunération . brut"
// lorsqu'elle n'incluera plus les frais professionnels.
export const rémunérationBruteDottedName = 'salarié . cotisations . assiette'
export const lodeomDottedName =
	'salarié . cotisations . exonérations . lodeom . montant'
export const heuresSupplémentairesDottedName =
	'salarié . temps de travail . heures supplémentaires'
export const heuresComplémentairesDottedName =
	'salarié . temps de travail . heures complémentaires'

export type Répartition = {
	IRC: number
	Urssaf: number
}

export type MonthState = {
	rémunérationBrute: number
	options: Options
	lodeom: {
		value: number
		répartition: Répartition
	}
	régularisation: {
		value: number
		répartition: Répartition
	}
}

export type Options = {
	heuresSupplémentaires: number
	heuresComplémentaires: number
	rémunérationETP: number
	rémunérationPrimes: number
	rémunérationHeuresSup: number
}

export type RégularisationMethod = 'annuelle' | 'progressive'

const getDateForContexte = (monthIndex: number, year: number): string => {
	const date = new Date(year, monthIndex)

	return date.toLocaleDateString('fr')
}

const getMonthlyLodeom = (
	date: string,
	rémunérationBrute: number,
	options: Options,
	engine: Engine<DottedName>
): number => {
	const lodeom = engine.evaluate({
		valeur: lodeomDottedName,
		unité: '€/mois',
		contexte: {
			date,
			[rémunérationBruteDottedName]: rémunérationBrute,
			[heuresSupplémentairesDottedName]: options.heuresSupplémentaires,
			[heuresComplémentairesDottedName]: options.heuresComplémentaires,
		},
	})

	return lodeom.nodeValue as number
}

const getTotalLodeom = (
	rémunérationBrute: number,
	SMIC: number,
	coefT: number,
	engine: Engine<DottedName>
): number => {
	const lodeom = engine.evaluate({
		valeur: lodeomDottedName,
		arrondi: 'non',
		contexte: {
			[rémunérationBruteDottedName]: rémunérationBrute,
			'salarié . temps de travail . SMIC': SMIC,
			'salarié . cotisations . exonérations . T': coefT,
		},
	})

	return lodeom.nodeValue as number
}

const emptyRépartition = {
	IRC: 0,
	Urssaf: 0,
}

const getRépartition = (
	rémunération: number,
	réduction: number,
	engine: Engine<DottedName>
): Répartition => {
	const contexte = {
		[rémunérationBruteDottedName]: rémunération,
		[lodeomDottedName]: réduction,
	}
	const IRC =
		(engine.evaluate({
			valeur: `${lodeomDottedName} . imputation retraite complémentaire`,
			unité: '€/mois',
			contexte,
		})?.nodeValue as number) ?? 0
	const Urssaf =
		(engine.evaluate({
			valeur: `${lodeomDottedName} . imputation sécurité sociale`,
			unité: '€/mois',
			contexte,
		})?.nodeValue as number) ?? 0

	return {
		IRC,
		Urssaf,
	}
}

export const getInitialLodeomMoisParMois = (
	year: number,
	engine: Engine<DottedName>
): MonthState[] => {
	const rémunérationBrute =
		(engine.evaluate({
			valeur: rémunérationBruteDottedName,
			arrondi: 'oui',
			unité: '€/mois',
		})?.nodeValue as number) ?? 0
	const heuresSupplémentaires =
		(engine.evaluate({
			valeur: heuresSupplémentairesDottedName,
			unité: 'heures/mois',
		})?.nodeValue as number) ?? 0
	const heuresComplémentaires =
		(engine.evaluate({
			valeur: heuresComplémentairesDottedName,
			unité: 'heures/mois',
		})?.nodeValue as number) ?? 0
	const rémunérationETP = 0
	const rémunérationPrimes = 0
	const rémunérationHeuresSup = 0

	if (!rémunérationBrute) {
		return Array(12).fill({
			rémunérationBrute,
			options: {
				heuresSupplémentaires,
				heuresComplémentaires,
				rémunérationETP,
				rémunérationPrimes,
				rémunérationHeuresSup,
			},
			lodeom: {
				value: 0,
				répartition: emptyRépartition,
			},
			régularisation: {
				value: 0,
				répartition: emptyRépartition,
			},
		}) as MonthState[]
	}

	return Array.from({ length: 12 }, (_item, monthIndex) => {
		const date = getDateForContexte(monthIndex, year)

		const lodeom = getMonthlyLodeom(
			date,
			rémunérationBrute,
			{
				heuresSupplémentaires,
				heuresComplémentaires,
				rémunérationETP,
				rémunérationPrimes,
				rémunérationHeuresSup,
			},
			engine
		)
		const répartition = getRépartition(rémunérationBrute, lodeom, engine)

		return {
			rémunérationBrute,
			options: {
				heuresSupplémentaires,
				heuresComplémentaires,
				rémunérationETP,
				rémunérationPrimes,
				rémunérationHeuresSup,
			},
			lodeom: {
				value: lodeom,
				répartition,
			},
			régularisation: {
				value: 0,
				répartition: emptyRépartition,
			},
		}
	})
}

export const reevaluateLodeomMoisParMois = (
	data: MonthState[],
	engine: Engine<DottedName>,
	year: number,
	régularisationMethod: RégularisationMethod
): MonthState[] => {
	const totalRémunérationBrute = sumAll(
		data.map((monthData) => monthData.rémunérationBrute)
	)

	if (!totalRémunérationBrute) {
		return data.map((monthData) => {
			return {
				...monthData,
				lodeom: {
					value: 0,
					répartition: emptyRépartition,
				},
				régularisation: {
					value: 0,
					répartition: emptyRépartition,
				},
			}
		})
	}

	const rémunérationBruteCumulées = getRémunérationBruteCumulées(data)
	const SMICCumulés = getSMICCumulés(data, year, engine)
	// Si on laisse l'engine calculer T dans le calcul de l'exonération Lodeom,
	// le résultat ne sera pas bon à cause de l'assiette de cotisations du contexte
	const coefT = engine.evaluate({
		valeur: 'salarié . cotisations . exonérations . T',
	}).nodeValue as number

	const reevaluatedData = data.reduce(
		(reevaluatedData: MonthState[], monthState, monthIndex) => {
			const rémunérationBrute = monthState.rémunérationBrute
			const options = monthState.options
			const lodeom = {
				value: 0,
				répartition: emptyRépartition,
			}
			const régularisation = {
				value: 0,
				répartition: emptyRépartition,
			}

			if (!rémunérationBrute) {
				return [
					...reevaluatedData,
					{
						rémunérationBrute,
						options,
						lodeom,
						régularisation,
					},
				]
			}

			if (régularisationMethod === 'progressive') {
				// La régularisation progressive du mois N est la différence entre l'exonération Lodeom
				// calculée pour la rémunération totale jusqu'à N (comparée au SMIC équivalent pour ces N mois)
				// et la somme des N-1 exonérations déjà accordées (en incluant les régularisations).
				const lodeomTotale = getTotalLodeom(
					rémunérationBruteCumulées[monthIndex],
					SMICCumulés[monthIndex],
					coefT,
					engine
				)
				const lodeomCumulée = sumAll(
					reevaluatedData.map(
						(monthData) =>
							monthData.lodeom.value + monthData.régularisation.value
					)
				)
				régularisation.value = lodeomTotale - lodeomCumulée

				if (régularisation.value > 0) {
					lodeom.value = régularisation.value
					lodeom.répartition = getRépartition(
						rémunérationBrute,
						lodeom.value,
						engine
					)
					régularisation.value = 0
				} else if (régularisation.value < 0) {
					régularisation.répartition = getRépartition(
						rémunérationBrute,
						régularisation.value,
						engine
					)
				}
			} else if (régularisationMethod === 'annuelle') {
				const date = getDateForContexte(monthIndex, year)
				lodeom.value = getMonthlyLodeom(
					date,
					rémunérationBrute,
					options,
					engine
				)

				if (monthIndex === data.length - 1) {
					// La régularisation annuelle est la différence entre l'exonération Lodeom calculée
					// pour la rémunération annuelle (comparée au SMIC annuel) et la somme des exonérations
					// Lodeom déjà accordées.
					const lodeomTotale = getTotalLodeom(
						rémunérationBruteCumulées[monthIndex],
						SMICCumulés[monthIndex],
						coefT,
						engine
					)
					const currentLodeomCumulée =
						lodeom.value +
						sumAll(reevaluatedData.map((monthData) => monthData.lodeom.value))
					régularisation.value = lodeomTotale - currentLodeomCumulée

					if (lodeom.value + régularisation.value > 0) {
						lodeom.value += régularisation.value
						lodeom.répartition = getRépartition(
							rémunérationBrute,
							lodeom.value,
							engine
						)
						régularisation.value = 0
					} else if (régularisation.value < 0) {
						régularisation.répartition = getRépartition(
							rémunérationBrute,
							régularisation.value,
							engine
						)
					}
				}
			}

			return [
				...reevaluatedData,
				{
					rémunérationBrute,
					options,
					lodeom,
					régularisation,
				},
			]
		},
		[]
	)

	return reevaluatedData
}

const getSMICCumulés = (
	data: MonthState[],
	year: number,
	engine: Engine<DottedName>
): number[] => {
	return data.reduce((SMICCumulés: number[], monthData, monthIndex) => {
		// S'il n'y a pas de rémunération ce mois-ci, il n'y a pas d'exonération Lodeom
		// et il ne faut pas compter le SMIC de ce mois-ci dans le SMIC cumulé.
		if (!monthData.rémunérationBrute) {
			SMICCumulés.push(0)

			return SMICCumulés
		}

		const contexte = {
			date: getDateForContexte(monthIndex, year),
		} as Situation

		if (!monthData.options.rémunérationETP) {
			contexte[heuresSupplémentairesDottedName] =
				monthData.options.heuresSupplémentaires
			contexte[heuresComplémentairesDottedName] =
				monthData.options.heuresComplémentaires
		}

		let SMIC = engine.evaluate({
			valeur: 'salarié . temps de travail . SMIC',
			unité: '€/mois',
			contexte,
		}).nodeValue as number

		if (monthData.options.rémunérationETP) {
			const SMICHoraire = engine.evaluate({
				valeur: 'SMIC . horaire',
				contexte,
			}).nodeValue as number
			// On retranche les primes et le paiements des heures supplémentaires à la rémunération versée
			// et on la compare à la rémunération équivalente "mois complet" sans les primes
			const prorata =
				(monthData.rémunérationBrute -
					monthData.options.rémunérationPrimes -
					monthData.options.rémunérationHeuresSup) /
				monthData.options.rémunérationETP
			// On applique ce prorata au SMIC mensuel et on y ajoute les heures supplémentaires et complémentaires
			SMIC =
				SMIC * prorata +
				SMICHoraire *
					(monthData.options.heuresSupplémentaires +
						monthData.options.heuresComplémentaires)
		}

		let SMICCumulé = SMIC
		if (monthIndex > 0) {
			// Il faut aller chercher la dernière valeur positive du SMIC cumulé.
			const previousSMICCumulé =
				SMICCumulés.findLast((SMICCumulé) => SMICCumulé > 0) ?? 0
			SMICCumulé = previousSMICCumulé + SMIC
		}

		SMICCumulés.push(SMICCumulé)

		return SMICCumulés
	}, [])
}

const getRémunérationBruteCumulées = (data: MonthState[]) => {
	return data.reduce(
		(rémunérationBruteCumulées: number[], monthData, monthIndex) => {
			// S'il n'y a pas de rémunération ce mois-ci, il n'y a pas d'exonération Lodeom
			// et elle ne compte pas non plus pour la régularisation des mois à venir.
			if (!monthData.rémunérationBrute) {
				rémunérationBruteCumulées.push(0)

				return rémunérationBruteCumulées
			}

			let rémunérationBruteCumulée = monthData.rémunérationBrute
			if (monthIndex > 0) {
				// Il faut aller chercher la dernière valeur positive de la rémunération cumulée.
				const previousRémunérationBruteCumulée =
					rémunérationBruteCumulées.findLast(
						(rémunérationBruteCumulée) => rémunérationBruteCumulée > 0
					) ?? 0
				rémunérationBruteCumulée =
					previousRémunérationBruteCumulée + monthData.rémunérationBrute
			}

			rémunérationBruteCumulées.push(rémunérationBruteCumulée)

			return rémunérationBruteCumulées
		},
		[]
	)
}
