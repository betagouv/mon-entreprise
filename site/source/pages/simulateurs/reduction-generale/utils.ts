import { sumAll } from 'effect/Number'
import { DottedName } from 'modele-social'
import Engine from 'publicodes'

import { Situation } from '@/domaine/Situation'

// TODO: remplacer "salarié . cotisations . assiette" par "salarié . rémunération . brut"
// lorsqu'elle n'incluera plus les frais professionnels.
export const rémunérationBruteDottedName = 'salarié . cotisations . assiette'
export const réductionGénéraleDottedName =
	'salarié . cotisations . exonérations . réduction générale'
export const heuresSupplémentairesDottedName =
	'salarié . temps de travail . heures supplémentaires'
export const heuresComplémentairesDottedName =
	'salarié . temps de travail . heures complémentaires'

export type Répartition = {
	IRC: number
	Urssaf: number
	chômage: number
}

export type MonthState = {
	rémunérationBrute: number
	options: Options
	réductionGénérale: {
		value: number
		répartition: Répartition
	}
	régularisation: {
		value: number
		répartition: Répartition
	}
}

export type Options = {
	heuresSupplémentaires: number
	heuresComplémentaires: number
	rémunérationETP: number
	rémunérationPrimes: number
	rémunérationHeuresSup: number
}

export type RégularisationMethod = 'annuelle' | 'progressive'

const getDateForContexte = (monthIndex: number, year: number): string => {
	const date = new Date(year, monthIndex)

	return date.toLocaleDateString('fr')
}

const getMonthlyRéductionGénérale = (
	date: string,
	rémunérationBrute: number,
	options: Options,
	engine: Engine<DottedName>
): number => {
	const réductionGénérale = engine.evaluate({
		valeur: réductionGénéraleDottedName,
		unité: '€/mois',
		contexte: {
			date,
			[rémunérationBruteDottedName]: rémunérationBrute,
			[heuresSupplémentairesDottedName]: options.heuresSupplémentaires,
			[heuresComplémentairesDottedName]: options.heuresComplémentaires,
		},
	})

	return réductionGénérale.nodeValue as number
}

const getTotalRéductionGénérale = (
	rémunérationBrute: number,
	SMIC: number,
	coefT: number,
	engine: Engine<DottedName>
): number => {
	const réductionGénérale = engine.evaluate({
		valeur: réductionGénéraleDottedName,
		arrondi: 'non',
		contexte: {
			[rémunérationBruteDottedName]: rémunérationBrute,
			'salarié . temps de travail . SMIC': SMIC,
			'salarié . cotisations . exonérations . T': coefT,
		},
	})

	return réductionGénérale.nodeValue as number
}

const emptyRépartition = {
	IRC: 0,
	Urssaf: 0,
	chômage: 0,
}

const getRépartition = (
	rémunération: number,
	réduction: number,
	engine: Engine<DottedName>
): Répartition => {
	const contexte = {
		[rémunérationBruteDottedName]: rémunération,
		[réductionGénéraleDottedName]: réduction,
	}
	const IRC =
		(engine.evaluate({
			valeur: `${réductionGénéraleDottedName} . imputation retraite complémentaire`,
			unité: '€/mois',
			contexte,
		})?.nodeValue as number) ?? 0
	const Urssaf =
		(engine.evaluate({
			valeur: `${réductionGénéraleDottedName} . imputation sécurité sociale`,
			unité: '€/mois',
			contexte,
		})?.nodeValue as number) ?? 0
	const chômage =
		(engine.evaluate({
			valeur: `${réductionGénéraleDottedName} . imputation chômage`,
			unité: '€/mois',
			contexte,
		})?.nodeValue as number) ?? 0

	return {
		IRC,
		Urssaf,
		chômage,
	}
}

export const getInitialRéductionGénéraleMoisParMois = (
	year: number,
	engine: Engine<DottedName>
): MonthState[] => {
	const rémunérationBrute =
		(engine.evaluate({
			valeur: rémunérationBruteDottedName,
			arrondi: 'oui',
			unité: '€/mois',
		})?.nodeValue as number) ?? 0
	const heuresSupplémentaires =
		(engine.evaluate({
			valeur: heuresSupplémentairesDottedName,
			unité: 'heures/mois',
		})?.nodeValue as number) ?? 0
	const heuresComplémentaires =
		(engine.evaluate({
			valeur: heuresComplémentairesDottedName,
			unité: 'heures/mois',
		})?.nodeValue as number) ?? 0
	const rémunérationETP = 0
	const rémunérationPrimes = 0
	const rémunérationHeuresSup = 0

	if (!rémunérationBrute) {
		return Array(12).fill({
			rémunérationBrute,
			options: {
				heuresSupplémentaires,
				heuresComplémentaires,
				rémunérationETP,
				rémunérationPrimes,
				rémunérationHeuresSup,
			},
			réductionGénérale: {
				value: 0,
				répartition: emptyRépartition,
			},
			régularisation: {
				value: 0,
				répartition: emptyRépartition,
			},
		}) as MonthState[]
	}

	return Array.from({ length: 12 }, (_item, monthIndex) => {
		const date = getDateForContexte(monthIndex, year)

		const réductionGénérale = getMonthlyRéductionGénérale(
			date,
			rémunérationBrute,
			{
				heuresSupplémentaires,
				heuresComplémentaires,
				rémunérationETP,
				rémunérationPrimes,
				rémunérationHeuresSup,
			},
			engine
		)
		const répartition = getRépartition(
			rémunérationBrute,
			réductionGénérale,
			engine
		)

		return {
			rémunérationBrute,
			options: {
				heuresSupplémentaires,
				heuresComplémentaires,
				rémunérationETP,
				rémunérationPrimes,
				rémunérationHeuresSup,
			},
			réductionGénérale: {
				value: réductionGénérale,
				répartition,
			},
			régularisation: {
				value: 0,
				répartition: emptyRépartition,
			},
		}
	})
}

export const reevaluateRéductionGénéraleMoisParMois = (
	data: MonthState[],
	engine: Engine<DottedName>,
	year: number,
	régularisationMethod: RégularisationMethod
): MonthState[] => {
	const totalRémunérationBrute = sumAll(
		data.map((monthData) => monthData.rémunérationBrute)
	)

	if (!totalRémunérationBrute) {
		return data.map((monthData) => {
			return {
				...monthData,
				réductionGénérale: {
					value: 0,
					répartition: emptyRépartition,
				},
				régularisation: {
					value: 0,
					répartition: emptyRépartition,
				},
			}
		})
	}

	const rémunérationBruteCumulées = getRémunérationBruteCumulées(data)
	const SMICCumulés = getSMICCumulés(data, year, engine)
	// Si on laisse l'engine calculer T dans le calcul de la réduction générale,
	// le résultat ne sera pas bon à cause de l'assiette de cotisations du contexte
	const coefT = engine.evaluate({
		valeur: 'salarié . cotisations . exonérations . T',
	}).nodeValue as number

	const reevaluatedData = data.reduce(
		(reevaluatedData: MonthState[], monthState, monthIndex) => {
			const rémunérationBrute = monthState.rémunérationBrute
			const options = monthState.options
			const réductionGénérale = {
				value: 0,
				répartition: emptyRépartition,
			}
			const régularisation = {
				value: 0,
				répartition: emptyRépartition,
			}

			if (!rémunérationBrute) {
				return [
					...reevaluatedData,
					{
						rémunérationBrute,
						options,
						réductionGénérale,
						régularisation,
					},
				]
			}

			if (régularisationMethod === 'progressive') {
				// La régularisation progressive du mois N est la différence entre la réduction générale
				// calculée pour la rémunération totale jusqu'à N (comparée au SMIC équivalent pour ces N mois)
				// et la somme des N-1 réductions générales déjà accordées (en incluant les régularisations).
				const réductionGénéraleTotale = getTotalRéductionGénérale(
					rémunérationBruteCumulées[monthIndex],
					SMICCumulés[monthIndex],
					coefT,
					engine
				)
				const réductionGénéraleCumulée = sumAll(
					reevaluatedData.map(
						(monthData) =>
							monthData.réductionGénérale.value + monthData.régularisation.value
					)
				)
				régularisation.value =
					réductionGénéraleTotale - réductionGénéraleCumulée

				if (régularisation.value > 0) {
					réductionGénérale.value = régularisation.value
					réductionGénérale.répartition = getRépartition(
						rémunérationBrute,
						réductionGénérale.value,
						engine
					)
					régularisation.value = 0
				} else if (régularisation.value < 0) {
					régularisation.répartition = getRépartition(
						rémunérationBrute,
						régularisation.value,
						engine
					)
				}
			} else if (régularisationMethod === 'annuelle') {
				const date = getDateForContexte(monthIndex, year)
				réductionGénérale.value = getMonthlyRéductionGénérale(
					date,
					rémunérationBrute,
					options,
					engine
				)

				if (monthIndex === data.length - 1) {
					// La régularisation annuelle est la différence entre la réduction générale calculée
					// pour la rémunération annuelle (comparée au SMIC annuel) et la somme des réductions
					// générales déjà accordées.
					const réductionGénéraleTotale = getTotalRéductionGénérale(
						rémunérationBruteCumulées[monthIndex],
						SMICCumulés[monthIndex],
						coefT,
						engine
					)
					const currentRéductionGénéraleCumulée =
						réductionGénérale.value +
						sumAll(
							reevaluatedData.map(
								(monthData) => monthData.réductionGénérale.value
							)
						)
					régularisation.value =
						réductionGénéraleTotale - currentRéductionGénéraleCumulée

					if (réductionGénérale.value + régularisation.value > 0) {
						réductionGénérale.value += régularisation.value
						réductionGénérale.répartition = getRépartition(
							rémunérationBrute,
							réductionGénérale.value,
							engine
						)
						régularisation.value = 0
					} else if (régularisation.value < 0) {
						régularisation.répartition = getRépartition(
							rémunérationBrute,
							régularisation.value,
							engine
						)
					}
				}
			}

			return [
				...reevaluatedData,
				{
					rémunérationBrute,
					options,
					réductionGénérale,
					régularisation,
				},
			]
		},
		[]
	)

	return reevaluatedData
}

const getSMICCumulés = (
	data: MonthState[],
	year: number,
	engine: Engine<DottedName>
): number[] => {
	return data.reduce((SMICCumulés: number[], monthData, monthIndex) => {
		// S'il n'y a pas de rémunération ce mois-ci, il n'y a pas de réduction générale
		// et il ne faut pas compter le SMIC de ce mois-ci dans le SMIC cumulé.
		if (!monthData.rémunérationBrute) {
			SMICCumulés.push(0)

			return SMICCumulés
		}

		const contexte = {
			date: getDateForContexte(monthIndex, year),
		} as Situation

		if (!monthData.options.rémunérationETP) {
			contexte[heuresSupplémentairesDottedName] =
				monthData.options.heuresSupplémentaires
			contexte[heuresComplémentairesDottedName] =
				monthData.options.heuresComplémentaires
		}

		let SMIC = engine.evaluate({
			valeur: 'salarié . temps de travail . SMIC',
			unité: '€/mois',
			contexte,
		}).nodeValue as number

		if (monthData.options.rémunérationETP) {
			const SMICHoraire = engine.evaluate({
				valeur: 'SMIC . horaire',
				contexte,
			}).nodeValue as number
			// On retranche les primes et le paiements des heures supplémentaires à la rémunération versée
			// et on la compare à la rémunération équivalente "mois complet" sans les primes
			const prorata =
				(monthData.rémunérationBrute -
					monthData.options.rémunérationPrimes -
					monthData.options.rémunérationHeuresSup) /
				monthData.options.rémunérationETP
			// On applique ce prorata au SMIC mensuel et on y ajoute les heures supplémentaires et complémentaires
			SMIC =
				SMIC * prorata +
				SMICHoraire *
					(monthData.options.heuresSupplémentaires +
						monthData.options.heuresComplémentaires)
		}

		let SMICCumulé = SMIC
		if (monthIndex > 0) {
			// Il faut aller chercher la dernière valeur positive du SMIC cumulé.
			const previousSMICCumulé =
				SMICCumulés.findLast((SMICCumulé) => SMICCumulé > 0) ?? 0
			SMICCumulé = previousSMICCumulé + SMIC
		}

		SMICCumulés.push(SMICCumulé)

		return SMICCumulés
	}, [])
}

const getRémunérationBruteCumulées = (data: MonthState[]) => {
	return data.reduce(
		(rémunérationBruteCumulées: number[], monthData, monthIndex) => {
			// S'il n'y a pas de rémunération ce mois-ci, il n'y a pas de réduction générale
			// et elle ne compte pas non plus pour la régularisation des mois à venir.
			if (!monthData.rémunérationBrute) {
				rémunérationBruteCumulées.push(0)

				return rémunérationBruteCumulées
			}

			let rémunérationBruteCumulée = monthData.rémunérationBrute
			if (monthIndex > 0) {
				// Il faut aller chercher la dernière valeur positive de la rémunération cumulée.
				const previousRémunérationBruteCumulée =
					rémunérationBruteCumulées.findLast(
						(rémunérationBruteCumulée) => rémunérationBruteCumulée > 0
					) ?? 0
				rémunérationBruteCumulée =
					previousRémunérationBruteCumulée + monthData.rémunérationBrute
			}

			rémunérationBruteCumulées.push(rémunérationBruteCumulée)

			return rémunérationBruteCumulées
		},
		[]
	)
}
