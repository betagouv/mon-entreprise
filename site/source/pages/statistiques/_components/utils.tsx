export function formatDay(date: string | Date, language: string) {
	return new Date(date).toLocaleString(language, {
		weekday: 'long',
		day: 'numeric',
		month: 'long',
	})
}
export function formatMonth(date: string | Date, language: string) {
	return new Date(date).toLocaleString(language, {
		month: 'long',
		year: 'numeric',
	})
}
export function formatIndicator(number: number, language: string) {
	return Intl.NumberFormat(language, {
		notation: 'compact',
	}).format(number)
}

export function formatProgression(
	previousValue: number,
	currentValue: number,
	language: string
) {
	return Intl.NumberFormat(language, {
		style: 'percent',
		signDisplay: 'exceptZero',
	}).format(currentValue / previousValue - 1)
}
export function emojiSatisfaction(satisfaction: number) {
	return satisfaction > 0.85
		? '😀'
		: satisfaction > 0.65
		? '🙂'
		: satisfaction > 0.5
		? '😐'
		: '😕'
}

export function messageTypeSatisfaction(satisfaction: number) {
	return satisfaction > 0.85
		? 'success'
		: satisfaction > 0.65
		? 'primary'
		: satisfaction > 0.5
		? 'info'
		: 'error'
}
