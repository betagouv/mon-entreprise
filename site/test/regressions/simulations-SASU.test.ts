import { it } from 'vitest'

import { configSASU } from '@/pages/Simulateurs/sasu/_simulationConfig'

import rémunérationSASUSituations from './simulations-SASU.yaml'
import { runSimulations } from './utils'

it('calculate assimilé salarié', () => {
	runSimulations(
		rémunérationSASUSituations,
		[
			...(configSASU['objectifs exclusifs'] ?? []),
			...(configSASU.objectifs ?? []),
		],
		configSASU.situation
	)
})
