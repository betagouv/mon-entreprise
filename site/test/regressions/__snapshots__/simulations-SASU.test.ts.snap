// Vitest Snapshot v1

exports[`calculate assimilé salarié > ACRE 1`] = `
"dirigeant . rémunération . totale: 10000
impôt . montant: 0
salarié . cotisations: 519
salarié . rémunération . net . payé après impôt: 596
salarié . rémunération . net . à payer avant impôt: 596"
`;

exports[`calculate assimilé salarié > ACRE 2`] = `
"dirigeant . rémunération . totale: 20000
impôt . montant: 0
salarié . cotisations: 1008
salarié . rémunération . net . payé après impôt: 1228
salarié . rémunération . net . à payer avant impôt: 1228"
`;

exports[`calculate assimilé salarié > ACRE 3`] = `
"dirigeant . rémunération . totale: 30000
impôt . montant: 873
salarié . cotisations: 1497
salarié . rémunération . net . payé après impôt: 1788
salarié . rémunération . net . à payer avant impôt: 1861"
`;

exports[`calculate assimilé salarié > JEI 1`] = `
"dirigeant . rémunération . totale: 48000
impôt . montant: 3247
salarié . cotisations: 1198
salarié . rémunération . net . payé après impôt: 2531
salarié . rémunération . net . à payer avant impôt: 2802"
`;

exports[`calculate assimilé salarié > avec charges 1`] = `
"dirigeant . rémunération . totale: 10000
impôt . montant: 0
salarié . cotisations: 396
salarié . rémunération . net . payé après impôt: 437
salarié . rémunération . net . à payer avant impôt: 437"
`;

exports[`calculate assimilé salarié > avec charges 2`] = `
"dirigeant . rémunération . totale: 20000
impôt . montant: 0
salarié . cotisations: 760
salarié . rémunération . net . payé après impôt: 907
salarié . rémunération . net . à payer avant impôt: 907"
`;

exports[`calculate assimilé salarié > échelle de rémunération 1`] = `
"dirigeant . rémunération . totale: 100
impôt . montant: undefined
salarié . cotisations: undefined
salarié . rémunération . net . payé après impôt: undefined
salarié . rémunération . net . à payer avant impôt: undefined"
`;

exports[`calculate assimilé salarié > échelle de rémunération 2`] = `
"dirigeant . rémunération . totale: 1000
impôt . montant: 0
salarié . cotisations: 70
salarié . rémunération . net . payé après impôt: 14
salarié . rémunération . net . à payer avant impôt: 14"
`;

exports[`calculate assimilé salarié > échelle de rémunération 3`] = `
"dirigeant . rémunération . totale: 2000
impôt . montant: 0
salarié . cotisations: 106
salarié . rémunération . net . payé après impôt: 61
salarié . rémunération . net . à payer avant impôt: 61"
`;

exports[`calculate assimilé salarié > échelle de rémunération 4`] = `
"dirigeant . rémunération . totale: 5000
impôt . montant: 0
salarié . cotisations: 215
salarié . rémunération . net . payé après impôt: 202
salarié . rémunération . net . à payer avant impôt: 202"
`;

exports[`calculate assimilé salarié > échelle de rémunération 5`] = `
"dirigeant . rémunération . totale: 10000
impôt . montant: 0
salarié . cotisations: 396
salarié . rémunération . net . payé après impôt: 437
salarié . rémunération . net . à payer avant impôt: 437"
`;

exports[`calculate assimilé salarié > échelle de rémunération 6`] = `
"dirigeant . rémunération . totale: 20000
impôt . montant: 0
salarié . cotisations: 760
salarié . rémunération . net . payé après impôt: 907
salarié . rémunération . net . à payer avant impôt: 907"
`;

exports[`calculate assimilé salarié > échelle de rémunération 7`] = `
"dirigeant . rémunération . totale: 50000
impôt . montant: 1704
salarié . cotisations: 1849
salarié . rémunération . net . payé après impôt: 2175
salarié . rémunération . net . à payer avant impôt: 2317"
`;

exports[`calculate assimilé salarié > échelle de rémunération 8`] = `
"dirigeant . rémunération . totale: 100000
impôt . montant: 9659
salarié . cotisations: 3622
salarié . rémunération . net . payé après impôt: 3907
salarié . rémunération . net . à payer avant impôt: 4712"
`;
